### PR TITLE
Follow up cleanups

### PR DESCRIPTION
**Pull request by lennoxho**
This should be my last PR for some time. Thanks for accepting my previous commits.

Includes:
1. Remove build dependency on `HAVE_PTHREAD`
2. Remove build dependency on `HAVE_STD_ALIGNED_VAL_T` and `ENABLE_ALIGNED_NEW_DELETE`
3. Remove redundant `tcmalloc.h` files & ensure there are no cross-build-tool references
4. Adopt Automake commit 26927d1 in the CMake build

Does not include:
1. Migrating from `__thread` to `thread_local` to take advantage of automatic object lifetimes
2. Remove `HAVE_STRUCT_MALLINFO`. Ran into couple of challenges:
    - `malloc.h` is a non-standard include
        - `__has_include` is one solution, but is not available on older compilers (and is technically C++17)
        - `!_WIN32` is risky, since `malloc.h` is non-standard (looks like it's a GLIBC extension)
    - `void_t`/`type_identity` SFINAE tricks only works for C++. `tcmalloc.h` appears to also cater to C users.
        - `!_WIN32` is risky, since `struct malloinfo` is non-standard
        - Checking for `__GLIBC_MINOR__` is probably a bad idea too, or are we assuming GLIBC for non-Windows?
-----

Testing done:

Autotools:
- WSL Ubuntu 22.04 LTS + GCC 13.1, ran unit tests
- MSYS2-MinGW [link](https://github.com/msys2/msys2-installer/releases/tag/2024-01-13) + GCC 13.2

VSProj:
- MSVC 19.00 (2015) + WinSDK 10.0.22621.0
- MSVC 19.38 (2022) + WinSDK 10.0.22621.0

CMake:
- WSL Ubuntu 22.04 LTS + GCC 13.1
- MSVC 19.38 + WinSDK 10.0.22621.0
- MinGW [link](https://github.com/brechtsanders/winlibs_mingw/releases/tag/13.2.0posix-17.0.5-11.0.1-ucrt-r3) + GCC 13.2

### CODE REVIEW

File: cmake/tcmalloc.h.in [hunk 1]
```diff
@@ -137,7 +141,7 @@ extern "C" {
   PERFTOOLS_DLL_DECL void tc_deletearray_nothrow(void* p,
                                                  const std::nothrow_t&) PERFTOOLS_NOTHROW;
 
-#if @HAVE_STD_ALIGN_VAL_T@ && __cplusplus >= 201703L
+#ifdef PERFTOOLS_HAVE_ALIGNED_NEW
```
Comment (lennoxho) [inline cmake/tcmalloc.h.in:144]: Note we will be losing a useful check here.

Currently the generated header file will match what is shipped in the tcmalloc binary.
With this change said check will be deferred to link time or even potentially startup time.

A solution would be to unconditionally export new variants of those functions by replacing `align_val_t` with our own type.
Then redefine the existing `align_val_t` functions based on those.

File: cmake/tcmalloc.h.in [hunk 2]
```diff
@@ -76,6 +74,12 @@
 # endif
 #endif
 
+#if defined(__cpp_aligned_new) || \
```
Comment (alk) [inline cmake/tcmalloc.h.in:77]: question: just cpp_aligned_new not enough?

File: cmake/tcmalloc.h.in (hunk 1 above)
Comment (alk) [reply cmake/tcmalloc.h.in:144]: Maybe I misunderstood, but I don't agree with playing games trying to mimic/re-implement align_val_t type.

But I agree with your first thesis that this commit is somewhat risky move.

**Comment by alk** [comment]
I made a pass over the code and it looks fine. Give me couple more days to think.

Also I just posted a RFC to mailing list about maybe upgrading to c++ 17 and eliminating our concerns above. See here: https://groups.google.com/g/gperftools/c/DO66YHh4tQ0/m/joRYMqp2CQAJ

File: cmake/tcmalloc.h.in (hunk 2 above)
Comment (lennoxho) [reply cmake/tcmalloc.h.in:77]: So there's some uncertainty here on MSVC's behaviour.

As you may know, MSVC will set `__cplusplus` to the wrong value unless [`/Zc:__cplusplus`](https://learn.microsoft.com/en-us/cpp/build/reference/zc-cplusplus) is also provided.
This is unfortunately still the default behaviour on the command line.

Now, I can't find an authoritative answer whether the C++ feature-test macros are also affected by this unfortunate situation.

That's why we have that additional `_MSVC_LANG` check, which serves the same role as `__cplusplus` - but never lies.

File: cmake/tcmalloc.h.in (hunk 1 above)
Comment (lennoxho) [reply cmake/tcmalloc.h.in:144]: I was basically suggesting
```
namespace tcmalloc { enum class align_val_t : std::size_t {}; }

PERFTOOLS_DLL_DECL void* tc_new_aligned2(size_t size, tcmalloc::align_val_t al);
...

#ifdef PERFTOOLS_HAVE_ALIGNED_NEW
  inline void* tc_new_aligned(size_t size, std::align_val_t al) {
      tc_new_aligned2(size,  tcmalloc::align_val_t((std::size_t)al));
  }
  ...
#endif
```

The idea is we make sure the aligned new/delete functions will always be available regardless of the producer's toolchain.

Downsides are - (1) We'll be repeating ourselves, (2) ABI break

**Comment by alk** [comment]
Applied. Thanks. For text time please consider writing commit message in more typical subject/new-line/body style.

**Pull request merged**